lib/express_validator.js

```diff
@@ -1,262 +1,335 @@
+  });
+
+  return function(req, res, next) {
+    var locations = ['body', 'params', 'query'];
+
+    req._validationErrors = [];
+    req.validationErrors = function(mapped) {
+      if (mapped && req._validationErrors.length > 0) {
+        var errors = {};
+        req._validationErrors.forEach(function(err) {
+          errors[err.param] = err;
+        });
+        return errors;
       }
-    });
-    Object.keys(_options.customSanitizers).forEach(function(customName) {
-      methods[customName] = function() {
-        var args = [value].concat(Array.prototype.slice.call(arguments));
-        var result = _options.customSanitizers[customName].apply(null, args);
-        request.updateParam(param, result);
-        return result;
+      return (req._validationErrors.length > 0) ? req._validationErrors : false;
+    };
+
+    locations.forEach(function(location) {
+      req['sanitize' + capitalCase(location)] = function(param) {
+        return new SanitizerChain(param, req, [location]);
       };
     });
-    return methods;
-  };
 
-  function checkParam(req, getter) {
-    return function(param, failMsg) {
-
-      var value;
-
-      // If param is not an array, then split by dot notation
-      // returning an array. It will return an array even if
-      // param doesn't have the dot notation.
-      //      'blogpost' = ['blogpost']
-      //      'login.username' = ['login', 'username']
-      // For regex matches you can access the parameters using numbers.
-      if (!Array.isArray(param)) {
-        param = typeof param === 'number' ?
-          [param] :
-          param.split('.').filter(function(e){
-            return e !== '';
-          });
-      }
+    req.sanitize = function(param) {
+      return new SanitizerChain(param, req, ['params', 'query', 'body']);
+    };
 
-      // Extract value from params
-      param.map(function(item) {
-        if (value === undefined) {
-          value = getter(item);
-        } else {
-          value = value[item];
-        }
-      });
-      param = param.join('.');
-
-      var errorHandler = function(msg) {
-        var error = _options.errorFormatter(param, msg, value);
-
-        if (req._validationErrors === undefined) {
-          req._validationErrors = [];
-        }
-        req._validationErrors.push(error);
-
-        if (req.onErrorCallback) {
-          req.onErrorCallback(msg);
-        }
-        return this;
+    locations.forEach(function(location) {
+      req['check' + capitalCase(location)] = function(param, failMsg) {
+        return new ValidatorChain(param, failMsg, req, location, options);
       };
+    });
+
+    req.checkFiles = function(param, failMsg) {
+      return new ValidatorChain(param, failMsg, req, 'files', options);
+    };
 
-      var methods = [];
+    req.checkHeaders = function(param, failMsg) {
+      if(param === 'referrer') {param = 'referer';}
```

Comment:
I'm just going to comment on this first instance, but throughout the refactor you have a lot of one-line conditionals. I realize now that I should have put a code style guide in this repo, but I would prefer:
1) A space between the "if" and opening parens
2) A space between the closing parens and the opening bracket
3) Conditionals to be multi-line (unless they are ternary operators)